File: Backend/models/doctor_work_schedule.py
from sqlalchemy import (
    Column,
    Integer,
    String,
    Text,
    Date,
    DateTime,
    ForeignKey
)
from sqlalchemy.sql import func
from db import Base


class DoctorWorkSchedule(Base):
    __tablename__ = "doctor_work_schedules"

    id = Column(Integer, primary_key=True, index=True)

    doctor_id = Column(
    Integer,
    ForeignKey("users.id"),
    nullable=True
)

    work_shift_id = Column(
        Integer,
        ForeignKey("work_shifts.id"),
        nullable=False
    )

    work_date = Column(
        Date,
        nullable=False
    )

    max_patients = Column(
        Integer,
        default=10
    )

    current_patients = Column(
        Integer,
        default=0
    )

    status = Column(
        String(20),
        default="available"
    )

    note = Column(Text)

    created_at = Column(
        DateTime(timezone=True),
        server_default=func.now()
    )

    updated_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now()
    )

File: flutter_application_1/lib/models/doctor_work_schedule.dart
class DoctorWorkSchedule {
  final String id;

  final String? doctorId;

  final String workShiftId;

  final DateTime workDate;

  final int maxPatients;

  final int currentPatients;

  final String status;

  final String? note;

  // join từ bảng doctors
  final String? doctorName;

  final String? doctorCode;

  final DateTime createdAt;

  final DateTime updatedAt;

  DoctorWorkSchedule({
    required this.id,
    this.doctorId,
    required this.workShiftId,
    required this.workDate,
    required this.maxPatients,
    required this.currentPatients,
    required this.status,
    this.note,
    this.doctorName,
    this.doctorCode,
    required this.createdAt,
    required this.updatedAt,
  });

  factory DoctorWorkSchedule.fromJson(Map<String, dynamic> json) {
    return DoctorWorkSchedule(
      id: json['id']?.toString() ?? '',

      doctorId: json['doctor_id']?.toString(),

      workShiftId: json['work_shift_id']?.toString() ?? '',

      workDate: DateTime.parse(json['work_date'].toString()),

      maxPatients: json['max_patients'] ?? 0,

      currentPatients: json['current_patients'] ?? 0,

      status: json['status'] ?? 'unknown',

      note: json['note']?.toString(),

      doctorName: json['doctor_name']?.toString(),

      doctorCode: json['doctor_code']?.toString(),

      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'].toString())
          : DateTime.now(),

      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'].toString())
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "doctor_id": doctorId,
      "work_shift_id": workShiftId,
      "work_date": workDate.toIso8601String().split('T')[0],
      "max_patients": maxPatients,
      "current_patients": currentPatients,
      "status": status,
      "note": note,
    };
  }

  DoctorWorkSchedule copyWith({
    String? id,
    String? doctorId,
    String? workShiftId,
    DateTime? workDate,
    int? maxPatients,
    int? currentPatients,
    String? status,
    String? note,
    String? doctorName,
    String? doctorCode,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return DoctorWorkSchedule(
      id: id ?? this.id,
      doctorId: doctorId ?? this.doctorId,
      workShiftId: workShiftId ?? this.workShiftId,
      workDate: workDate ?? this.workDate,
      maxPatients: maxPatients ?? this.maxPatients,
      currentPatients: currentPatients ?? this.currentPatients,
      status: status ?? this.status,
      note: note ?? this.note,
      doctorName: doctorName ?? this.doctorName,
      doctorCode: doctorCode ?? this.doctorCode,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
